[gui2] Shortcut manager + shortcuts for selecting nodes

somebody1234 opened the pull request:
### Pull Request Description
- Closes #7874
  - Chosen solution: Use the dashboard's shortcut manager to register keybinds and match mouse shortcuts
  - Switch existing shortcuts to shortcut manager as well
  - Switch visualizations to use the shortcut manager too
- Closes #7865
  - Implements all shortcuts for node selection, *except* the arrow key shortcut
- Adds circle cursor selection brush
  - The fade out from inactivity (which is present in the Rust source code) is currently missing

Other changes:
- Change everything to use `pointerdown` as appropriate, to ensure the event handlers are called in the right order

### Important Notes
The arrow key shortcut to navigate to the nearest node in that direction *has not* been implemented.
This may need discussion on how it should work

The implementation in general is pretty inefficient - this is intentional, performance can be optimized in the future

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

somebody1234 committed:
Prepare to allow adding new shortcuts from GUI

somebody1234 committed:
wip

somebody1234 committed:
Add basic shortcut handling; add node selected state; fix minor bugs in shortcut registry

somebody1234 committed:
Add custom cursor and selection brush

somebody1234 committed:
wip: selections

somebody1234 committed:
wip: update selection type based on mouse event modifiers

somebody1234 committed:
Fix cursor not moving when panning

somebody1234 committed:
Finish implementing selection

somebody1234 committed:
Fix incorrect node rect size

somebody1234 committed:
Fix toggle behavior

somebody1234 committed:
Fix single node selection

somebody1234 committed:
Fix imports

somebody1234 committed:
Replace svg cursor with HTML div cursor

somebody1234 committed:
Make the `shortcutRegistry` prop optional

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Add existing keyboard shortcuts to registry; switch cursor and brush to scene coordinates

Frizi commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -35,46 +41,12 @@ function updateExprRect(id: ExprId, rect: Rect) {
 }
 
 useWindowEvent('keydown', (e) => {
   if (keyboardBusy()) return
```
This whole handler can be removed. The `keyboardBusy` check was only here to **not** run the key handler when having any element focused. That being said, is the shortcut manager also handling that case?

somebody1234 replied:
oops - yeah, meant to remove this after further testing. but after the refactor, this should no longer be relevant. currently `keyboardBusy` *is* being used for the post-refactor handlers

Frizi commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -106,12 +262,25 @@ function moveNode(id: ExprId, delta: Vec2) {
         @delete="graphStore.deleteNode(id)"
         @updateExprRect="updateExprRect"
         @updateContent="(range, c) => updateNodeContent(id, range, c)"
         @movePosition="moveNode(id, $event)"
```
The `movePosition` handler should handle the case of selected nodes in following way:
- When moved node is in the set of currently selected nodes, move all selected nodes.
- Otherwise move only the node that was interacted with.

Note that the event already contains a delta (the dragged amount in scene coordinates), so it's just a matter of adding a check and a loop.

Frizi commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -106,12 +262,25 @@ function moveNode(id: ExprId, delta: Vec2) {
         @updateContent="(range, c) => updateNodeContent(id, range, c)"
         @movePosition="moveNode(id, $event)"
       />
+      <SelectionBrush
+        v-if="selection.dragging && selectionStart && selectionSize"
+        :start="selectionStart"
+        :size="selectionSize"
+      />
+      <CustomCursor
+        v-else-if="navigator.sceneMousePos"
+        :x="navigator.sceneMousePos.x"
+        :y="navigator.sceneMousePos.y"
+      />
```
Combine `SelectionBrush` and `CustomCursor` into a single cursor-drawing component that handles both. That way we will be able to add nice transitions between selected and non-selected states (see how it behaves in Rust gui for reference). The animation itself doesn't need to be added now, but can be if it ends up being an insignificant amount of work (e.g. just a css transition does the trick).

somebody1234 replied:
animation has been added locally. it's probably not 100% identical to the rust one timing wise, but i think it's good enough for now.
(also, i'm not sure where the rust animation is implemented)

Frizi commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -137,6 +306,7 @@ function moveNode(id: ExprId, delta: Vec2) {
   position: relative;
   contain: layout;
   overflow: clip;
+  cursor: none;
```
There are some other cursor styles already present inside the node component. Make sure to remove those.

Frizi commented on app/gui2/src/components/GraphNode.vue:
```diff
@@ -244,11 +248,19 @@ onUpdated(() => {
-    e.stopPropagation()
+// TODO [sb]: restore delete functionality (DELETE key)
+function handleClick(event: PointerEvent) {
+  if (shortcutRegistry.matchesMouseAction('replace-nodes-selection', event)) {
+    emit('replaceSelection')
+  } else if (shortcutRegistry.matchesMouseAction('add-to-nodes-selection', event)) {
+    emit('update:selected', true)
+  } else if (shortcutRegistry.matchesMouseAction('remove-from-nodes-selection', event)) {
+    emit('update:selected', false)
+  } else if (
+    shortcutRegistry.matchesMouseAction('toggle-nodes-selection', event) ||
+    shortcutRegistry.matchesMouseAction('invert-nodes-selection', event)
+  ) {
+    emit('update:selected', !props.selected)
```
The whole shortcut management looks messy and highly repetitive right now. I think it's worth giving it another shot with a better API in mind. Something like this could be a nice goal:

```ts
// Definition of keybinds namespace. Executed during module load (top level of script).
// The name (first argument) is used only for config matching purposes.
const graphBindings = keybinds.defineNamespace('graph-editor', {
    // `Mod` is a custom modifier, matching either `Ctrl` on Windows or `Command` on Mac
    undo: ['Mod+Z'],
    // Multiple alternative shortcuts can be specified, both work the same way.
    redo: ['Mod+Y', 'Mod+Shift+Z'],
    // Allows binding to mouse (pointer) buttons, accepts modifier keys with them.
    dragScene: ['PointerAux', 'Ctrl+PointerMain'],
    // ...more actions. The order here doesn't matter, but names do.
})
  
  
// 'keydown' event handler of a component:
function onKeyDown(event: KeyboardEvent) {
    graphBindings.handle(event, {
        // The order of specified object keys determines the order of matching. Only the first
        // matching action is executed. Once an action is matched, the event automatically stops
        // propagating and default behavior is prevented.
        undo() {
            // Undo action matched here
        },
        redo() {
            // Redo action matched here
        },
    
        [keybinds.DefaultHandler]() {
            // Optionally, when no action was matched in this block, do something else
        },
    })
}

// 'mousedown' event handler in a component. Should also work the same with pointer events.
function onKeyDown(event: MouseEvent) {
    // perform singular match. Equivalent to `handle` call with a single action, returns `true` if
    // the event is matching, automatically stops propagation and prevents default behavior.
    if (graphBindings.match(event, 'dragScene')) {

    }
}
```

Note, we could totally integrate that into the `usePointer` composable, so it accepts the event name to react to.

somebody1234 committed:
Merge cursor and brush; add animation to brush

somebody1234 committed:
Initial implementation of new shortcut registry

somebody1234 committed:
Merge branch 'develop' into wip/sb/shortcut-manager

somebody1234 committed:
Fix `GraphNode` styling

somebody1234 committed:
Revert `ide-desktop` changes

somebody1234 committed:
Cleanup `defineKeybinds`

somebody1234 commented:
status update:
- the new shortcut management API is more or less complete. it's very significantly different from the React one though, so all changes to the React one have been reverted.
  - however there are quite a few bugs related to events being triggered in the wrong situations.
  - note that the `match` API has not been implemented. it shouldn't be too hard, but i'm not sure it's that useful currently.
- the other unresolved issues are not yet addressed.

somebody1234 committed:
Switch `GraphNode` to use new mouse handler

somebody1234 committed:
Move handlers to end of script

somebody1234 committed:
Add state for latest selected node

somebody1234 committed:
Display latest selected node over other nodes

somebody1234 committed:
Fixes

somebody1234 committed:
Change `movePosition` to be selection-aware

Frizi commented on app/gui2/src/util/shortcuts.ts:
```diff
@@ -0,0 +1,400 @@
+  }
+}
+
+function parseKeybindString(string: string): Keybind | Mousebind {
```
This could have a few unit tests. Those can be added at the end of this file inside an `if (import.meta.vitest)` block.

Frizi commented on app/gui2/src/util/shortcuts.ts:
```diff
@@ -0,0 +1,400 @@
+// === Constants ===
+// =================
+
+// FIXME [sb]: This should be present in the keybinds too, just like `Mod`.
+/** The key known as the `Delete` key for the current platform. */
+const delete_ = isMacLike ? 'Backspace' : 'Delete'
```
I agree, and I think it should be implemented here. Not necessarily as a modifier, but as a key. You could add an extra entry to `allKeys`, and handle it specially during parsing, similarily to '+'. You could potentially add an additional mapping object for use there, similar to modifiers. Just doesn't have to include all keys, only ones needing an override (e.g. `+`, `OsDelete` key).

Frizi commented on app/gui2/src/components/GraphNode.vue:
```diff
@@ -363,77 +334,120 @@ function updatePreprocessor(module: string, method: string, ...args: string[]) {
-          :offset="0"
-          @updateExprRect="updateExprRect"
-        />
+  <Teleport :to="floatingNodeContainerNode" :disabled="!isLatestSelected">
```
I don't like the use of `Teleport` here. The node reordering on selection should instead be handled differently, and should maintain correct ordering that matches whole selection history. I can think of two ways to do it:

- You could add an ever-growing z-index to a node every time it is selected.
- Alternatively, we could directly reorder the nodes on the `GraphEditor`/`graphStore` level by reinserting the node to the nodes map, effectively moving it to the end (the `Map` iteration order is guaranteed to match insertion order).

somebody1234 committed:
Fix node layering

somebody1234 committed:
Add unit tests for keybind string parsing

somebody1234 committed:
Fix definition of `osDelete`

somebody1234 committed:
Merge keyboard and mouse handlers

somebody1234 committed:
Fix selection brush scaling

somebody1234 committed:
Refactor `usePointer` to work with `mouseHandler`s

somebody1234 committed:
Refactor viz builtins module

somebody1234 committed:
Switch viz to new shortcut manager; fix viz bugs

somebody1234 committed:
Hide circle cursor when system cursor is visible; fix viz resizing

somebody1234 committed:
Fix viz selector disappearing

somebody1234 committed:
Hide cursor on visualization buttons

somebody1234 committed:
Merge branch 'develop' into wip/sb/shortcut-manager

somebody1234 commented:
changes since last review, for convenience:
https://github.com/enso-org/enso/pull/7893/files/94ab8a826c62c907ba20141344cb981090b22ec7..40bdce08be599fb782c53745d93bf783f7b5a212?w=1

<blockquote><img src="https://avatars.githubusercontent.com/u/4046547?s=400&v=4" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso/pull/7893">[gui2] Shortcut manager + shortcuts for selecting nodes by somebody1234 · Pull Request #7893 · enso-org/enso</a></strong></div><div>Pull Request Description

Closes #7874

Chosen solution: Use the dashboard&#39;s shortcut manager to register keybinds and match mouse shortcuts
Switch existing shortcuts to shortcut manager as wel...</div></blockquote>

somebody1234 commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -269,22 +271,23 @@ const mouseHandler = nodeBindings.mouseHandler(
-        :floating-node-container-node="floatingNodeContainerNode"
-        @update:selected="setSelected(id, $event), $event && (latestSelectedNode = id)"
-        @replaceSelection="selectedNodes.clear(), selectedNodes.add(id), (latestSelectedNode = id)"
+        @update:selected="setSelected(id, $event), $event && updateLatestSelectedNode(id)"
+        @replaceSelection="
+          selectedNodes.clear(), selectedNodes.add(id), updateLatestSelectedNode(id)
+        "
```
these are pretty funky... on one hand, i guess the right thing to do is to extract them to a function
on the other hand i don't think it's a bad thing to have the definitions closer to (or rather, exactly at) the place where they are used, *if* the definitions are simple enough

somebody1234 committed:
Change test message

somebody1234 commented:
i think the parts that need the most attention would be the changes in `@/util/*`

Frizi committed:
fix import in template

somebody1234 commented:
@Frizi linting says there's an issue with invalid modifiers btw. not sure what's the best solution for that... i *think* it's required because both `useNavigator` and `usePointer` bind to the same events. thoughts?

somebody1234 commented:
i might've gone overboard on the lint fixes...

Frizi committed:
improve node design, add local d3, lint

Frizi committed:
cleanup type patches

Frizi committed:
use props in templates consistently

Frizi committed:
reenable eslint rule
